src/core/dbus-execute.c

```diff
@@ -1849,6 +1850,47 @@ int bus_exec_context_set_transient_property(
 
                 return 1;
 
+        } else if (streq(name, "UnsetEnvironment")) {
+
+                _cleanup_strv_free_ char **l = NULL, **q = NULL;
+
+                r = sd_bus_message_read_strv(message, &l);
+                if (r < 0)
+                        return r;
+
+                if (!strv_env_name_or_assignment_is_valid(l))
+                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid UnsetEnvironment= list.");
+
+                r = unit_full_printf_strv(u, l, &q);
+                if (r < 0)
+                        return r;
+
+                if (mode != UNIT_CHECK) {
+                        if (strv_length(q) == 0) {
+                                c->unset_environment = strv_free(c->unset_environment);
+                                unit_write_drop_in_private_format(u, mode, name, "UnsetEnvironment=");
+                        } else {
+                                _cleanup_free_ char *joined = NULL;
+                                char **e;
+
+                                e = strv_env_merge(2, c->unset_environment, q);
+                                if (!e)
+                                        return -ENOMEM;
+
+                                strv_free(c->unset_environment);
+                                c->unset_environment = e;
+
+                                /* We write just the new settings out to file, with unresolved specifiers */
```

Comment:
`UnsetEnvironment` doesn't appear to resolve specifiers, so the comment should probably be removed as well as `unit_full_printf_strv`.

On the other hand, `Environment` allow using specifiers, so if `UnsetEnvironment` is supposed to be used to unset everything set by `Environment`, it should support specifiers too.

Reply:
I have modified the test a bit to show what I mean.
```diff
diff --git a/test/test-execute/exec-unset-environment.service b/test/test-execute/exec-unset-environment.service
index b439739a6..42a966158 100644
--- a/test/test-execute/exec-unset-environment.service
+++ b/test/test-execute/exec-unset-environment.service
@@ -2,7 +2,7 @@
 Description=Test for UnsetEnvironment

 [Service]
-ExecStart=/bin/sh -x -c 'test "$$FOO" = "bar" && test "$$QUUX" = "" && test "$$VAR3" = "value3" && test "$$VAR4" = "" && test "$$VAR5" = "value5"'
+ExecStart=/bin/sh -x -c 'test "$$FOO" = "bar" && test "$$QUUX" = "" && test "$$VAR3" = "value3" && test "$$VAR4" = "" && test "$$VAR5" = "value5" && test "$$%u" = ""'
 Type=oneshot
-Environment=FOO=bar QUUX=waldo VAR3=value3 VAR4=value4 VAR5=value5
-UnsetEnvironment=QUUX=waldo VAR3=somethingelse VAR4
+Environment=FOO=bar QUUX=waldo VAR3=value3 VAR4=value4 VAR5=value5 %u=%u
+UnsetEnvironment=QUUX=waldo VAR3=somethingelse VAR4 %u
```
It currently fails, because `%u` is not unset